### PR TITLE
Feature/add nypl support

### PR DESCRIPTION
**Pull request by winkidney**
Hi,
I add image download support for NYPL (New York Public Library) at https://digitalcollections.nypl.org

Example url:  https://digitalcollections.nypl.org/items/6bee9930-73a0-0134-38bf-00505686a51c

Would you like to have a review and merge it into the project?

Feel free to point out any issue if necessary (I wrote the code according to iipimage's code).

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -42,6 +42,7 @@ log = "0.4"
 env_logger = "0.7"
 url = "2"
 fixedbitset = "0.3"
+json = "0.12.4"
```
Comment (lovasoa) [inline Cargo.toml:45]: Can we just use serde_json as in the other dezoomers ? This gives simpler code, better json structure validation, better error handling, and better error messages.

File: src/nypl/mod.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+/// A dezoomer for NYPL images
+#[derive(Default)]
+pub struct NYPLImage;
+
+const NYPL_IMAGE_VIEW_PREFIX: &str = "https://digitalcollections.nypl.org/items/";
+const NYPL_META_PREFIX: &str = "https://access.nypl.org/image.php/";
+const NYPL_META_POSTFIX: &str = "/tiles/config.js";
+
+fn _get_image_url_from_meta_url(meta_url: &str) -> String {
```
Comment (lovasoa) [inline src/nypl/mod.rs:19]: If this function is used, its name shouldn't start with an underscore

File: src/nypl/mod.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    meta_url.replace(NYPL_META_PREFIX, "")
+        .replace(NYPL_META_POSTFIX, "")
+}
+
+macro_rules! nypl_meta_format_tpl {
+    () => {
+        "https://access.nypl.org/image.php/{}/tiles/config.js"
+    }
+}
```
Comment (lovasoa) [inline src/nypl/mod.rs:28]: It's a little bit overkill. You can either inline that or use NYPL_META_PREFIX and NYPL_META_POSTFIX

File: src/nypl/mod.rs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+impl Dezoomer for NYPLImage {
+    fn name(&self) -> &'static str { "NYPLImage" }
+    fn zoom_levels(&mut self, data: &DezoomerInput) -> Result<ZoomLevels, DezoomerError> {
+        if data.uri.starts_with(NYPL_IMAGE_VIEW_PREFIX) {
+            self.assert(data.uri.contains(NYPL_IMAGE_VIEW_PREFIX))?;
```
Comment (lovasoa) [inline src/nypl/mod.rs:34]: Isn't the assert redundant with the if ?

File: src/nypl/mod.rs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            let image_id = data.uri.replace(NYPL_IMAGE_VIEW_PREFIX, "");
```
Comment (lovasoa) [inline src/nypl/mod.rs:35]: Don't we want to parse the uri completely ? What if the user entered `https://digitalcollections.nypl.org/items/a14f3200-fac1-012f-f7a4-58d385a7bbd0#item-data` ?

File: src/nypl/mod.rs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NYPLImage")
+    }
+}
+
+impl TilesRect for Level {
+    fn size(&self) -> Vec2d {
+        let reverse_level = self.metadata.levels - self.level - 1;
+        self.metadata.size / 2_u32.pow(reverse_level)
+    }
```
Comment (lovasoa) [inline src/nypl/mod.rs:79]: Did you check how they do the rounding ? Many formats use `ceil(size / 2.^levels)`, or have more complex sizing algorithm, that just happen to match this `floor(size / 2.^levels)` on most but not all image sizes.

File: src/nypl/mod.rs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+    fn tile_size(&self) -> Vec2d { self.metadata.tile_size }
+
+    fn tile_url(&self, Vec2d { x, y }: Vec2d) -> String {
+        let Vec2d { x: _width, .. } = self.size().ceil_div(self.tile_size());
```
Comment (lovasoa) [inline src/nypl/mod.rs:84]:
```suggestion
```

File: src/nypl/mod.rs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        let width = meta["size"]["width"].as_str().unwrap()
+            .parse::<u32>().unwrap();
+        let height = meta["size"]["height"].as_str().unwrap()
+            .parse::<u32>().unwrap();
+        let _tile_width = meta["tilesize"].as_str().unwrap()
+            .parse::<u32>().unwrap();
+        let size = Vec2d { x: width, y: height };
+        let tile_size = Vec2d{x: _tile_width, y: _tile_width};
+        let levels= 1;
+        Ok(Metadata { size, tile_size, levels })
```
Comment (lovasoa) [inline src/nypl/mod.rs]: We can remove that once we use serde_json

File: src/nypl/mod.rs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    fn size(&self) -> Vec2d {
+        let reverse_level = self.metadata.levels - self.level - 1;
+        self.metadata.size / 2_u32.pow(reverse_level)
+    }
+
+    fn tile_size(&self) -> Vec2d { self.metadata.tile_size }
+
+    fn tile_url(&self, Vec2d { x, y }: Vec2d) -> String {
+        let Vec2d { x: _width, .. } = self.size().ceil_div(self.tile_size());
+        format!("https://access.nypl.org/image.php/{id}/tiles/0/12/{x}_{y}.png",
```
Comment (lovasoa) [inline src/nypl/mod.rs:85]: I see the `.png` is hardcoded, but the meta-information file contains `"format":"png"`. Shouldn't the format be dynamic ?

File: README.md [hunk 10]
```diff
@@ -136,6 +137,12 @@ for panoramas, virtual tours, photoshperes, and other 3d zoomable images.
 dezoomify-rs supports downloading individual image planes from such images.
 You need to provide the xml meta-information file for the image.
 
+### NYPLImage
+
+[NYPLIMage](https://digitalcollections.nypl.org) is just a tile-set of png-image
```
Comment (lovasoa) [inline README.md:142]: Maybe let the user know that they can give either the image viewer URL or the config.js URL to dezoomify-rs.

**Review by lovasoa — CHANGES_REQUESTED** [review]
This looks good ! Thank you for your contribution ! I noted a few changes to make, and then I'll merge :)

**Comment by winkidney** [comment]
Thanks for your quick reply.
The issue will be updated once I have time.

File: Cargo.toml (hunk 1 above)
Comment (winkidney) [reply Cargo.toml:45]: will be done.

File: src/nypl/mod.rs (hunk 2 above)
Comment (winkidney) [reply src/nypl/mod.rs:19]: Will be done.

File: src/nypl/mod.rs (hunk 3 above)
Comment (winkidney) [reply src/nypl/mod.rs:28]: Seems reasonable

File: src/nypl/mod.rs (hunk 5 above)
Comment (winkidney) [reply src/nypl/mod.rs:35]: Currently we simply parse it without any tag or query string (like ? and #).
Maybe complex parse is reasonable? If you have strong opinion on it, I will handle more cases.

Comment (lovasoa) [reply src/nypl/mod.rs:35]: maybe just a small regex like : `https?://digitalcollections.nypl.org/items/([a-f0-9\-]+)` ?

File: src/nypl/mod.rs (hunk 6 above)
Comment (winkidney) [reply src/nypl/mod.rs:79]: I haven't done fully test about that.
I've use the code to download an image and it works well, I got the image correctly : )

Comment (winkidney) [reply src/nypl/mod.rs:79]: image

File: src/nypl/mod.rs (hunk 9 above)
Comment (winkidney) [reply src/nypl/mod.rs:85]: It seems that the whole site use png as the default zoom image format.
Anyway, I will change the code to use dynamic extension while possible.

File: src/nypl/mod.rs (hunk 8 above)
Comment (winkidney) [reply src/nypl/mod.rs]: I use the same logic as the old one.
I prefer to not build a strut to parse it (because it will not be used at other place).

**Comment by winkidney** [comment]
@lovasoa I've updated the PR. Please have new review for me, thanks!
And, I wounder how the CI error occured:
```
/Users/runner/.cargo/bin/cargo build
   Compiling dezoomify-rs v2.5.2 (/Users/runner/work/dezoomify-rs/dezoomify-rs)
error[E0463]: can't find crate for `pin_project_internal` which `futures` depends on
Error:  --> src/lib.rs:6:5
  |
6 | use futures::stream::StreamExt;
  |     ^^^^^^^ can't find crate
```
Could you please help to figure it out for me?

**Comment by lovasoa** [comment]
Thanks for the update ! We are almost ready to merge. We just need to fix the json parsing.

For the CI error, I'll look into it, it doesn't seem to come from your code.

**Comment by lovasoa** [comment]
I updated the dependencies and this seems to have fixed CI. The latest CI error is due to your code.

File: src/nypl/mod.rs [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        return Some(cap[1].to_owned());
+    }
+    return None;
+}
+
+impl Dezoomer for NYPLImage {
+    fn name(&self) -> &'static str { "NYPLImage" }
+    fn zoom_levels(&mut self, data: &DezoomerInput) -> Result<ZoomLevels, DezoomerError> {
+        if data.uri.starts_with(NYPL_IMAGE_VIEW_PREFIX) {
+            let image_id = parse_image_id(data.uri.as_str()).unwrap();
```
Comment (lovasoa) [inline src/nypl/mod.rs:38]: We probably don't want to crash if the image id is not found

**Review by lovasoa — CHANGES_REQUESTED** [review]

**Comment by lovasoa** [comment]
Ok, I'll take take it from there, I'll merge this today

**Comment by winkidney** [comment]
Thanks for your improvement and fix!

**Comment by lovasoa** [comment]
After a closer look, it looks like the site uses OpenSeadragon with the dzi file format internally. They just use a custom meta-information format and a custom `getTileUrl` function for custom tile URLs. But the tile positions and level size calculations are those of the dzi format. In particular, we cannot simply ignore the `overlap` parameter.

Here is a screenshot of a dezoomed image with the current dezoomer

image

Since the overlap is not taken into account, we end up with duplicated pixel rows and columns around tile edges. 

<details>

```js
    e.prototype.initUuid = function (e, t) {
      return null == this.configs[e] ? $.ajax(this.options.tileBase + '/' + e + '/tiles/config.js?callback=?', {
        dataType: 'jsonp',
        cache: !0,
        timeout: 5000,
        success: function (i) {
          return function (n) {
            var o,
            r,
            s,
            a,
            l;
            if (null == n) {
              if (null != t.url) return s = new Image,
              s.onload = function () {
                var n;
                return n = {
                },
                n[0] = {
                  type: 'legacy-image-pyramid',
                  levels: [
                    {
                      height: s.height,
                      width: s.width,
                      url: t.url,
                      mimetype: 'image/jpeg'
                    }
                  ]
                },
                i.configs[e] = n,
                null != t && null != t.success ? t.success(n)  : void 0
              },
              s.src = t.url
            } else {
              r = {
              },
              a = n.configs;
              for (l in a) o = a[l],
              r[l] = {
                type: 'dzi',
                height: parseInt(o.size.height),
                width: parseInt(o.size.width),
                tileSize: parseInt(o.tilesize),
                tileOverlap: parseInt(o.overlap),
                getTileUrl: function (t, n, o) {
                  return i.options.tileBase + '/' + e + '/tiles/0/' + t + '/' + n + '_' + o + '.png'
                }
              };
              if (i.configs[e] = r, null != t && null != t.success) return t.success(r)
            }
          }
        }(this),
        error: function (e) {
          return function () {
            return null != t && null != t.error ? t.error()  : void 0
          }
        }(this)
      })  : null != t && null != t.success ? t.success(this.configs[e])  : void 0
    }
``` 

</details>

**Review by lovasoa — APPROVED** [review]

**Pull request merged**